lib/SILOptimizer/Differentiation/PullbackCloner.cpp

```diff
@@ -2994,13 +3093,70 @@ AdjointValue PullbackCloner::Implementation::accumulateAdjointsDirect(
       }
       return makeAggregateAdjointValue(lhsVal->getType(), newElements);
     }
+    // x + (baseAdjoint, index, eltToAdd)
+    // There are 2 possible cases -
```

Comment:
The whole intention of new adjoint kind is to prevent extraneous materialization of adjoints. Here it seems you're materializing them. Also, why do you need these explicit cases? Why can't you simply call `accumulateAdointsDirect(base, rhs)`, wrap result into `AddElement(, eltToAdd)` and return it? All that zero folding, etc. will be already handled for you.

Reply:
I see the issue here and I see what you were saying earlier. I was trying to constrain AddElement adjoint's base adjoint for no reason, it seems. It looks like it can be anything, which should truly let us defer materialization until `materializationAdjoint` is called.